juju/juju
# Correct Constraint Validation for Moving Subnets

## Description of change

This patch changes the way we validate space constraint integrity for moving subnets.

The first version considered application connectivity in aggregate, which was incorrect. We now consider unique sets of connectivity for units across an application that will be affected by the move.

New functionality to perform the interrogation is in _integrity.go_, and new convenience types for collections of subnets have been added to _core/network_.

## QA steps

This QA uses MicroStack. We need 3 subnets to test properly, so follow the instructions [here](https://discourse.juju.is/t/openstack-multi-space-support/2672), repeating the subnet addition step to add 192.168.224.0/24.

- Bootstrap to the MicroStack.
- `juju add-space beta 192.168.223.0/24`.
- `juju deploy cs:~jameinel/ubuntu-lite-7 --constraints spaces=beta`

`juju show-machine 0` will show that the machine is connected to the subnet above. Now deploy another unit of the application to a machine connected to 2 subnets.

- `juju add-space gamma 192.168.224.0/24`.
- `juju add-machine --constraints "spaces=beta,gamma"`
- `juju add-unit ubuntu-lite --to 1`

Again, `juju show-machine 1` will show connection to 2 subnets. Now move the subnet in `gamma` to `beta`. This should succeed, because both machines remain connected to `beta` and the application constraint is satisfied.

- `juju move-to-space beta 192.168.224.0/24`.

Now try to move the only subnet on machine 0 to gamma. This should fail, because ubuntu/0 would violate the application space constraint.

- `juju move-to-space gamma 192.168.223.0/24`.
- Error message should read:
```
ERROR cannot update space "gamma": moving subnet(s) to space "gamma" violates space constraints for application "ubuntu-lite": beta
        units not connected to the space: ubuntu-lite/0
```

## Documentation changes

None.

## Bug reference

N/A
